File: java/Standard.java
public enum Standard implements Coup {
    Pierre,
    Papier,
    Ciseaux;

    public boolean gagneContre(Standard autre) {
        return
            (this == Pierre && autre == Ciseaux ||
            this == Ciseaux && autre == Papier ||
            this == Papier && autre == Pierre);
    }

    /* on ne peut pas redéfinir compare : une enum hérite de java.lang.Enum
       qui implémente déjà Comparable et elle est marquée final. */
}

File: ocaml/standard.ml
type coup = Pierre | Papier | Ciseaux

let coups = [Pierre; Papier; Ciseaux]
            
let to_string x =
  match x with
  | Pierre -> "pierre"
  | Papier -> "papier"
  | Ciseaux -> "ciseaux"

let of_string x =
  match x with
  | "pierre" -> Pierre
  | "papier" -> Papier
  | "ciseaux" -> Ciseaux
  | _ -> invalid_arg "Standard.of_string"

let (>>) a b =
  match a, b with
  | Pierre, Ciseaux
  | Ciseaux, Papier
  | Papier, Pierre -> true
  | _ -> false
